File: hashing/C++/pointsOnStr8Line.cpp
#include<bits/stdc++.h>
int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }

int Solution::maxPoints(vector<int> &A, vector<int> &B) {
        int n = A.size();
        if (n <= 1) return n;

        int maxPoints = 0;

        for (int i = 0; i < n; ++i) {
            unordered_map<string, int> slopes;
            int duplicates = 1;
            
            for (int j = i + 1; j < n; ++j) {
                if (A[i] == A[j] && B[i] == B[j]) {
                    ++duplicates;
                } else {
                    int dx = A[j] - A[i];
                    int dy = B[j] - B[i];
                    int g = gcd(dx, dy);
                    
                    // Normalize slope
                    dx /= g;
                    dy /= g;
                    
                    // Ensure the direction is consistent
                    if (dx < 0) {
                        dx = -dx;
                        dy = -dy;
                    }
                    
                    if (dx == 0) {
                        dy = 1; // Vertical line
                    }
                    
                    string slope = to_string(dx) + "," + to_string(dy);
                    slopes[slope]++;
                }
            }

            int currentMax = duplicates;
            for (const auto& pair : slopes) {
                currentMax = max(currentMax, pair.second + duplicates);
            }

            maxPoints = max(maxPoints, currentMax);
        }

        return maxPoints;
    }

File: hashing/java/PointsOnStr8Line.java
public class Solution {
    public int maxPoints(ArrayList<Integer> A, ArrayList<Integer> B) {
        int n=A.size();
        int ans=2;
        if(n==0||n==1||n==2)return n;
        
        HashMap<Double, Integer> M= new HashMap<>();
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(i==j)continue;
                int x=A.get(i); int y=B.get(i);
                double slope= B.get(j)-y;
                int den=A.get(j)-x;
                
                if(den==0)slope= Double.MAX_VALUE;
                else slope=slope/den;
                
                M.put(slope, M.getOrDefault(slope, 0)+1);
                ans=Math.max(ans, M.get(slope)+1);
            }
            M.clear();
        }
        return ans;
    }
}